discord/bot.py

```diff
@@ -270,81 +270,172 @@ async def get_desynced_commands(
+                    return rtn
+                else:
+                    return self.callback(local, remote)
+
+        class DefaultSetComparison(DefaultComparison):
+            def check(self, local, remote) -> bool:
+                try:
+                    local = set(local)
+                except TypeError:
+                    pass
+                try:
+                    remote = set(remote)
+                except TypeError:
+                    pass
+                return super().check(local, remote)
+
+        type NestedComparison = dict[str, NestedComparison | DefaultComparison]
+
+        def _compare_defaults(
+            obj: Mapping[str, Any] | Any,
+            match: Mapping[str, Any] | Any,
+            schema: NestedComparison,
+        ) -> bool:
+            if not isinstance(match, Mapping) or not isinstance(obj, Mapping):
+                return obj != match
+            for field, comparison in schema.items():
+                remote = match.get(field, MISSING)
+                local = obj.get(field, MISSING)
+                if isinstance(comparison, dict):
+                    _compare_defaults(local, remote, comparison)
+                elif isinstance(comparison, DefaultComparison):
+                    if comparison.check(local, remote):
+                        return True
+            return False
 
         def _check_command(cmd: ApplicationCommand, match: Mapping[str, Any]) -> bool:
+            cmd = cmd.to_dict()
+
+            option_default_values = ([], MISSING)
+
+            def _option_comparison_check(local, remote) -> bool:
+                matching = (local in option_default_values) + (
+                    remote in option_default_values
+                )
+                if matching == 2:
+                    return False
+                elif matching == 1:
+                    return True
+                else:
+                    return len(local) != len(remote) or any(
+                        [
+                            _compare_defaults(local[x], remote[x], option_defaults)
+                            for x in range(len(local))
+                        ]
+                    )
+
+            choices_default_values = ([], MISSING)
+
+            def _choices_comparison_check(local, remote) -> bool:
+                matching = (local in choices_default_values) + (
+                    remote in choices_default_values
+                )
+                if matching == 2:
+                    return False
+                elif matching == 1:
+                    return True
+                else:
+                    return len(local) != len(remote) or any(
+                        [
+                            _compare_defaults(local[x], remote[x], choices_defaults)
+                            for x in range(len(local))
+                        ]
+                    )
+
+            defaults: NestedComparison = {
```

Comment:
This most likely can be made a CONSTANT and doesn't need to be redefined at every function invocation